### PR TITLE
[feature] Wordbreak and simple sequence predicate support

### PR DESCRIPTION
**Pull request by Ikcelaks**
This PR brings proper support for predicates, including `nonalpha`, which functions well as a word break.

Currently, the code appears to function as intended when running the tester in `test_ascii_string` mode, but there are still issues with running the rules test (namely, the tester tries to convert predicate triecodes to keycodes, which can't work. I have not yet tested on a keyboard.

The intention is to provide user replaceable funding for the common predicates (alpha, upper case alpha, digit, terminating punctuation, non terminating punctuation, any punctuation, and wildcard (any)), and also allow the user to specify their own custom predicates. The second is not yet implemented.

Predicates are tested by running a comparison function on the triecodes from the cursor. If the cursor is on a key that triggered an action, and the cursor is not doing the virtual output, the predicate is run on the last triecodes in the completion as a proxy (or the triecode of the actual key, if the completion is empty).

The format currently required in the generator config is a disaster, and I'll rework it before merge.

I will write up a better description tomorrow.

### CODE REVIEW

File: predicates.c [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+__attribute__((weak)) bool st_pred_upper_alpha(uint8_t triecode)
+{
+    const bool res = triecode >= 'A' && triecode <= 'Z';
+    st_debug(ST_DBG_SEQ_MATCH, " st_pred_upper_alpha: %d; Code: %#04X\n", res, triecode);
+    return res;
+}
+//////////////////////////////////////////////////////////////////////
+__attribute__((weak)) bool st_pred_alpha(uint8_t triecode)
+{
+    const bool res = triecode >= 'A' && triecode <= 'Z' && triecode >= 'a' && triecode <= 'z';
```
Comment (kamih) [inline predicates.c]: should be || instead of && no?

File: predicates.c [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    st_pred_terminating_punct,
+    st_pred_nonterminating_punct,
+    st_pred_punct,
+    st_pred_nonalpha,
+    st_pred_any
+};
+//////////////////////////////////////////////////////////////////////
+bool st_predicate_test_triecode(uint8_t predicate_index, uint8_t triecode)
+{
+    if (predicate_index > ST_PREDICATE_COUNT) {
```
Comment (kamih) [inline predicates.c]: `>=` ?

File: tester/Makefile [hunk 3]
```diff
@@ -33,8 +33,8 @@ CFLAGS := -Werror -Wundef \
 	-DST_TESTER \
 	-DSEQUENCE_TRANSFORM_DEBUG=1 \
 	-DSEQUENCE_TRANSFORM_ENHANCED_BACKSPACE=1 \
-	-DSEQUENCE_TRANSFORM_RULE_SEARCH=1 \
-	-DSEQUENCE_TRANSFORM_FALLBACK_BUFFER=1 \
+	-DSEQUENCE_TRANSFORM_RULE_SEARCH=0 \
+	-DSEQUENCE_TRANSFORM_FALLBACK_BUFFER=0 \
```
Comment (kamih) [inline tester/Makefile:37]: Tester needs these enabled.

File: generator/sequence_transform_data.py [hunk 4]
```diff
@@ -627,11 +622,11 @@ def generate_sequence_transform_data(data_header_file, test_header_file):
 
     for sequence, transform in seq_tranform_list:
         # Don't add rules with transformation functions to test header for now
-        if transform[-1] not in output_func_symbol_map:
-            transform_without_leading_wordbreak = transform[1:] if sequence.startswith(WORDBREAK_SYMBOL) else transform
-            test_rule = create_test_rule_c_string(symbol_map, sequence, transform_without_leading_wordbreak)
-            test_rule_c_sequences.append(test_rule[0])
-            test_rule_c_transforms.append(test_rule[1])
+        if len(transform) == 0 or transform[-1] not in output_func_symbol_map:
```
Comment (kamih) [inline generator/sequence_transform_data.py]: I know you don't like rule_search testing, but this should be `!= 0` :D

File: predicates.c (hunk 1 above)
Comment (Ikcelaks) [reply predicates.c]: Yep!

File: predicates.c (hunk 2 above)
Comment (Ikcelaks) [reply predicates.c]: Yep!

File: tester/Makefile (hunk 3 above)
Comment (Ikcelaks) [reply tester/Makefile:37]: I need them disabled for testing this branch at the moment, and it's much easier to switch branches if these settings are committed.

I'll roll this back at the end when this is no longer a draft

File: generator/sequence_transform_data.py (hunk 4 above)
Comment (Ikcelaks) [reply generator/sequence_transform_data.py]: Oh, hah! That was obviously unintentional. Good catch :)

**Comment by Ikcelaks** [comment]
I'm closing this PR, because it has become clear that explicit rule chaining is a prerequisite to getting consistent results from rules with Sequence MetaCharacters.

The issue bellow is intractable with the current matching scheme for chained rules that rematches previous matches (possibly with different results)
```
./tester -s ".*t"
--- str[0]: . ---
buffer: |.| (1)
output: |.| (1)
--- str[1]: * ---
buffer: |.👆| (2)
output: |.\| (2)  // `.👆 -> .\ ` correctly preempts `_👆 -> the`
--- str[2]: t ---
buffer: |.👆t| (3)
output: |.at| (3)``` // `_👆t -> _that` matches and incorrectly assumes that `_👆 -> the` was matched previously

**Comment by Ikcelaks** [comment]
I temporarily reopened this PR to fix the back merge, so that this code will be easier to reference when I reimplement it.

**Pull request closed without merging**